Index: rtw88/bf.c
===================================================================
--- rtw88.orig/bf.c
+++ rtw88/bf.c
@@ -75,7 +75,7 @@ void rtw_bf_assoc(struct rtw_dev *rtwdev
 		ether_addr_copy(bfee->mac_addr, bssid);
 		bfee->role = RTW_BFEE_MU;
 		bfee->p_aid = (bssid[5] << 1) | (bssid[4] >> 7);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 		bfee->aid = vif->cfg.aid;
 #else
 		bfee->aid = bss_conf->aid;
Index: rtw88/fw.c
===================================================================
--- rtw88.orig/fw.c
+++ rtw88/fw.c
@@ -196,7 +196,7 @@ legacy:
 	si->ra_report.desc_rate = rate;
 	si->ra_report.bit_rate = bit_rate;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)) || defined (OPENWRT)
 	sta->deflink.agg.max_rc_amsdu_len = get_max_amsdu_len(bit_rate);
 #else
 	sta->max_rc_amsdu_len = get_max_amsdu_len(bit_rate);
@@ -1257,7 +1257,7 @@ static struct sk_buff *rtw_get_rsvd_page
 
 	switch (rsvd_pkt->type) {
 	case RSVD_BEACON:
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 		skb_new = ieee80211_beacon_get_tim(hw, vif, &tim_offset, NULL, 0);
 #else
 		skb_new = ieee80211_beacon_get_tim(hw, vif, &tim_offset, NULL);
@@ -1272,7 +1272,7 @@ static struct sk_buff *rtw_get_rsvd_page
 		break;
 	case RSVD_NULL:
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 17)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)) || defined (OPENWRT)
 		skb_new = ieee80211_nullfunc_get(hw, vif, -1, false);
 #else
 		skb_new = ieee80211_nullfunc_get(hw, vif, false);
@@ -1283,7 +1283,7 @@ static struct sk_buff *rtw_get_rsvd_page
 		break;
 	case RSVD_QOS_NULL:
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 17)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)) || defined (OPENWRT)
 		skb_new = ieee80211_nullfunc_get(hw, vif, -1, true);
 #else
 		skb_new = ieee80211_nullfunc_get(hw, vif, true);
Index: rtw88/mac80211.c
===================================================================
--- rtw88.orig/mac80211.c
+++ rtw88/mac80211.c
@@ -64,7 +64,7 @@ static int rtw_ops_start(struct ieee8021
 	return ret;
 }
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0))
+#if ((LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0))) && !defined (OPENWRT)
 static void rtw_ops_stop(struct ieee80211_hw *hw)
 #else
 static void rtw_ops_stop(struct ieee80211_hw *hw, bool suspend)
@@ -378,7 +378,7 @@ static void rtw_conf_tx(struct rtw_dev *
 static void rtw_ops_bss_info_changed(struct ieee80211_hw *hw,
 				     struct ieee80211_vif *vif,
 				     struct ieee80211_bss_conf *conf,
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 				     u64 changed)
 #else
 				     u32 changed)
@@ -396,7 +396,7 @@ static void rtw_ops_bss_info_changed(str
 
 	if (changed & BSS_CHANGED_ASSOC) {
 		rtw_vif_assoc_changed(rtwvif, conf);
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 		if (vif->cfg.assoc) {
 #else
 		if (conf->assoc) {
@@ -405,7 +405,7 @@ static void rtw_ops_bss_info_changed(str
 			rtw_fw_download_rsvd_page(rtwdev);
 			rtw_send_rsvd_page_h2c(rtwdev);
 			rtw_fw_default_port(rtwdev, rtwvif);
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 			rtw_coex_media_status_notify(rtwdev, vif->cfg.assoc);
 #else
 			rtw_coex_media_status_notify(rtwdev, conf->assoc);
@@ -476,7 +476,7 @@ static void rtw_ops_bss_info_changed(str
 	mutex_unlock(&rtwdev->mutex);
 }
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 static int rtw_ops_start_ap(struct ieee80211_hw *hw,
 			    struct ieee80211_vif *vif,
 			    struct ieee80211_bss_conf *link_conf)
@@ -497,7 +497,7 @@ static int rtw_ops_start_ap(struct ieee8
 	return 0;
 }
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 static void rtw_ops_stop_ap(struct ieee80211_hw *hw,
 			    struct ieee80211_vif *vif,
 			    struct ieee80211_bss_conf *link_conf)
@@ -518,7 +518,7 @@ static void rtw_ops_stop_ap(struct ieee8
 
 static int rtw_ops_conf_tx(struct ieee80211_hw *hw,
 			   struct ieee80211_vif *vif,
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 			   unsigned int link_id, u16 ac,
 #else
 			   u16 ac,
@@ -990,7 +990,7 @@ static int rtw_ops_set_sar_specs(struct
 
 static void rtw_ops_sta_rc_update(struct ieee80211_hw *hw,
 				  struct ieee80211_vif *vif,
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 13, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 13, 0))
 				  struct ieee80211_sta *sta, u32 changed)
 #else
 				  struct ieee80211_link_sta *link_sta,
@@ -1008,7 +1008,7 @@ static void rtw_ops_sta_rc_update(struct
 }
 
 const struct ieee80211_ops rtw_ops = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)) || defined (OPENWRT)
 	.add_chanctx = ieee80211_emulate_add_chanctx,
 	.remove_chanctx = ieee80211_emulate_remove_chanctx,
 	.change_chanctx = ieee80211_emulate_change_chanctx,
Index: rtw88/main.c
===================================================================
--- rtw88.orig/main.c
+++ rtw88/main.c
@@ -194,7 +194,7 @@ static void rtw_vif_watch_dog_iter(void
 	struct rtw_vif *rtwvif = (struct rtw_vif *)vif->drv_priv;
 
 	if (vif->type == NL80211_IFTYPE_STATION)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 		if (vif->cfg.assoc)
 #else
 		if (vif->bss_conf.assoc)
@@ -588,7 +588,7 @@ EXPORT_SYMBOL(rtw_dump_reg);
 void rtw_vif_assoc_changed(struct rtw_vif *rtwvif,
 			   struct ieee80211_bss_conf *conf)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 	struct ieee80211_vif *vif = NULL;
 
 	if (conf)
@@ -1790,7 +1790,7 @@ static void rtw_vif_smps_iter(void *data
 {
 	struct rtw_dev *rtwdev = (struct rtw_dev *)data;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 	if (vif->type != NL80211_IFTYPE_STATION || !vif->cfg.assoc)
 #else
 	if (vif->type != NL80211_IFTYPE_STATION || !vif->bss_conf.assoc)
@@ -1798,13 +1798,13 @@ static void rtw_vif_smps_iter(void *data
 		return;
 
 	if (rtwdev->hal.txrx_1ss)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 		ieee80211_request_smps(vif, 0, IEEE80211_SMPS_STATIC);
 #else
 		ieee80211_request_smps(vif, IEEE80211_SMPS_STATIC);
 #endif
 	else
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 		ieee80211_request_smps(vif, 0, IEEE80211_SMPS_OFF);
 #else
 		ieee80211_request_smps(vif, IEEE80211_SMPS_OFF);
@@ -2311,7 +2311,7 @@ void rtw_core_deinit(struct rtw_dev *rtw
 		release_firmware(wow_fw->firmware);
 
 	destroy_workqueue(rtwdev->tx_wq);
-# if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 2, 0)
+# if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 2, 0)) || defined (OPENWRT)
 	timer_delete_sync(&rtwdev->tx_report.purge_timer);
 # else
 	del_timer_sync(&rtwdev->tx_report.purge_timer);
@@ -2542,7 +2542,7 @@ static void rtw_check_sta_active_iter(vo
 	if (vif->type != NL80211_IFTYPE_STATION)
 		return;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0)) || defined (OPENWRT)
 	if (vif->cfg.assoc || !is_zero_ether_addr(rtwvif->bssid))
 #else
 	if (vif->bss_conf.assoc || !is_zero_ether_addr(rtwvif->bssid))
Index: rtw88/main.h
===================================================================
--- rtw88.orig/main.h
+++ rtw88/main.h
@@ -36,7 +36,9 @@
 #define RHEL_RELEASE_VERSION(a, b) a<<8 & b
 #endif
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
+#define OPENWRT
+
+#if ((LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))) && !defined (OPENWRT)
 /**
  * abs_diff - return absolute value of the difference between the arguments
  * @a: the first argument
@@ -57,7 +59,7 @@
 })
 #endif
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)) && !defined (OPENWRT)
 /**
  * cfg80211_ssid_eq - compare two SSIDs
  * @a: first SSID
@@ -2418,7 +2420,7 @@ bool rtw_core_check_sta_active(struct rt
 void rtw_core_enable_beacon(struct rtw_dev *rtwdev, bool enable);
 void rtw_set_ampdu_factor(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 			  struct ieee80211_bss_conf *bss_conf);
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)) && !defined (OPENWRT)
 static inline bool rtw_ssid_equal(struct cfg80211_ssid *a,
                                  struct cfg80211_ssid *b)
 {
Index: rtw88/pci.c
===================================================================
--- rtw88.orig/pci.c
+++ rtw88/pci.c
@@ -1617,7 +1617,7 @@ static const struct rtw_hci_ops rtw_pci_
 
 static int rtw_pci_request_irq(struct rtw_dev *rtwdev, struct pci_dev *pdev)
 {
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0))) || defined (OPENWRT)
 	unsigned int flags = PCI_IRQ_INTX;
 #else
 	unsigned int flags = PCI_IRQ_LEGACY;
@@ -1694,7 +1694,7 @@ static int rtw_pci_napi_init(struct rtw_
 {
 	struct rtw_pci *rtwpci = (struct rtw_pci *)rtwdev->priv;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)) || defined (OPENWRT)
 	rtwpci->netdev = alloc_netdev_dummy(0);
 	if (!rtwpci->netdev)
 		return -ENOMEM;
@@ -1708,7 +1708,7 @@ static int rtw_pci_napi_init(struct rtw_
 #endif
 		, &rtwpci->napi
 		, rtw_pci_napi_poll
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)) && !defined (OPENWRT)
 		, NAPI_POLL_WEIGHT
 #endif
 	);
@@ -1721,7 +1721,7 @@ static void rtw_pci_napi_deinit(struct r
 
 	rtw_pci_napi_stop(rtwdev);
 	netif_napi_del(&rtwpci->napi);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)) || defined (OPENWRT)
 	free_netdev(rtwpci->netdev);
 #endif
 }
Index: rtw88/pci.h
===================================================================
--- rtw88.orig/pci.h
+++ rtw88/pci.h
@@ -215,7 +215,7 @@ struct rtw_pci {
 	bool running;
 
 	/* napi structure */
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)) || defined (OPENWRT)
 	struct net_device *netdev;
 #else
 	struct net_device netdev;
Index: rtw88/ps.c
===================================================================
--- rtw88.orig/ps.c
+++ rtw88/ps.c
@@ -334,7 +334,7 @@ void rtw_recalc_lps(struct rtw_dev *rtwd
 		__rtw_vif_recalc_lps(&data, new_vif);
 	rtw_iterate_vifs(rtwdev, rtw_vif_recalc_lps_iter, &data);
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))) || defined (OPENWRT)
 	if (data.count == 1 && data.found_vif->cfg.ps) {
 		rtwdev->ps_enabled = true;
 	} else {
Index: rtw88/rtw8703b.c
===================================================================
--- rtw88.orig/rtw8703b.c
+++ rtw88/rtw8703b.c
@@ -586,7 +586,7 @@ static void rtw8703b_pwrtrack_init(struc
 	dm_info->txagc_remnant_ofdm[RF_PATH_A] = 0;
 }
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)) && !defined (OPENWRT)
 #define FIELD_PREP_CONST FIELD_PREP
 #endif
 
Index: rtw88/rx.c
===================================================================
--- rtw88.orig/rx.c
+++ rtw88/rx.c
@@ -182,7 +182,7 @@ void rtw_update_rx_freq_from_ie(struct r
 	else
 		goto fill_rx_status;
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0))
+#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0))) || defined (OPENWRT)
 	channel_number = cfg80211_get_ies_channel_number(variable, ielen,
 							 NL80211_BAND_2GHZ);
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 17, 0))
